File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public static int maxProfit(int[] prices) {
        if(prices.length==1) return 0;
        if(prices.length==2) return Math.max(prices[1]-prices[0], 0);

        int left = 0;
        int right = 1;
        int tempProfit = 0;

        while(right<prices.length-1){
            if(prices[left+1]<prices[left]) {
                left++;
            }
            else if(prices[right]<prices[left]){
                left = right;
            }
            tempProfit = Math.max(prices[right] - prices[left], tempProfit);
            right++;
        }
        tempProfit = Math.max(prices[right] - prices[left], tempProfit);

        return tempProfit;
    }
}

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        best_profit = 0
        if len(prices) == 1:
            return 0
        left = 0 
        right = left + 1

        while right < len(prices) and left < right:
            new_profit = prices[right] - prices[left]
            if new_profit < 0:
                left = right
                right += 1
            elif new_profit > best_profit:
                best_profit = new_profit
                right += 1
            else:
                right +=1
        return best_profit
        